clang/lib/3C/DeclRewriter.cpp

```diff
@@ -252,199 +215,228 @@ void DeclRewriter::rewrite(RSet &ToRewrite) {
+  // When there are multiple levels of nested TagDecls, we need to process
+  // all the children of a TagDecl TD before TD itself so that (1) the
+  // definitions of the children end up before the definition of TD (since the
+  // rewriter preserves order of insertions) and (2) the definitions of the
+  // children have been removed from the body of TD before we read the body of
+  // TD to move it. In effect, we want to process the TagDecls in postorder.
+  // The easiest way to achieve this is to process them in order of their _end_
+  // locations.
+  std::sort(TagDeclsToDenest.begin(), TagDeclsToDenest.end(),
+            [&](TagDecl *TD1, TagDecl *TD2) {
+              return A.getSourceManager().isBeforeInTranslationUnit(
+                  TD1->getEndLoc(), TD2->getEndLoc());
+            });
+  for (TagDecl *TD : TagDeclsToDenest) {
+    // rewriteMultiDecl replaced the final "}" in the original source range with
+    // "};\n", so the new content of the source range should include the ";\n",
+    // which is what we want here. Except the rewriter has a bug where it
+    // adjusts the token range to include the final token _after_ mapping the
+    // offset to account for previous edits (it should be before). We work
+    // around the bug by adjusting the token range before calling the rewriter
+    // at all.
+    CharSourceRange CSR = Lexer::makeFileCharRange(
+        CharSourceRange::getTokenRange(TD->getSourceRange()), R.getSourceMgr(),
+        R.getLangOpts());
+    std::string DefinitionStr = R.getRewrittenText(CSR);
+    // Delete the definition from the old location.
+    rewriteSourceRange(R, CSR, "");
+    // We want to find the nearest ancestor DeclContext of TD that is _not_ a
+    // TagDecl and make TD a child of that DeclContext (named `Parent` below),
+    // just before the child `TopTagDecl` of `Parent` of which TD was originally
+    // a descendant.
+    //
+    // As of this writing, it seems that if TD is named, we get
+    // `Parent == TD->getDeclContext()` due to the code at
+    // https://github.com/correctcomputation/checkedc-clang/blob/fd4d8af4383d40af10ee8bc92b7bf88061a11035/clang/lib/Sema/SemaDecl.cpp#L16980-L16981,
+    // But that code doesn't run if TD is unnamed (which makes some sense
+    // because name visibility isn't an issue for TagDecls that have no name),
+    // and we want to de-nest TagDecls with names we assigned just like ones
+    // that were originally named, so we can't just use `TD->getDeclContext()`.
+    // In any event, maybe we wouldn't want to rely on this kind of internal
+    // Clang behavior.
+    DeclContext *TopTagDecl = TD;
+    for (;;) {
+      DeclContext *Parent = TopTagDecl->getLexicalParent();
+      if (!isa<TagDecl>(Parent))
+        break;
+      TopTagDecl = Parent;
+    }
+    // TODO: Use a wrapper like rewriteSourceRange that tries harder with
+    // macros, reports failure, etc.
+    R.InsertText(cast<Decl>(TopTagDecl)->getBeginLoc(), DefinitionStr);
```

Comment:
If you're interested in writing the `InsertText` wrapper, I'll definitely leave it to you to write it the way you want it.  Otherwise I can do it later.

Reply:
FTR: We decided to defer this to a follow-up issue #739, which is now cited by the TODO comment.